src/Tmds.Ssh/RemoteProcess.cs

```diff
@@ -13,83 +15,476 @@ public class RemoteProcess : IDisposable
+                {
+                    if (TryFindNewline(inspected: 0, out line))
+                    {
+                        return (0, line);
+                    }
+                    _mayHaveNewline = false;
+                }
+
+                int bytesDecoded = 0;
+                line = null;
+
+                if (sequence != null)
+                {
+                    ReadOnlySequence<byte> roSequence = sequence.AsReadOnlySequence();
+                    if (roSequence.IsSingleSegment)
+                    {
+                        if (TryDecodeAndFindNewLine(roSequence.FirstSpan, ref bytesDecoded, out line))
+                        {
+                            return (bytesDecoded, line);
+                        }
+                    }
+                    else
+                    {
+                        foreach (var segment in roSequence)
+                        {
+                            if (TryDecodeAndFindNewLine(segment.Span, ref bytesDecoded, out line))
+                            {
+                                return (bytesDecoded, line);
+                            }
+                        }
+                    }
+                }
+
+                if (eof)
+                {
+                    if (_length > 0)
+                    {
+                        Span<char> lineSpan = _buffer.AsSpan(_offset, _length);
+                        if (_stripNewline && lineSpan[0] == '\n')
+                        {
+                            lineSpan = lineSpan.Slice(1);
+                        }
+                        line = lineSpan.ToString();
+                        _length = 0;
+                    }
+                }
+
+                return (bytesDecoded, line);
+            }
+
+            private bool TryDecodeAndFindNewLine(ReadOnlySpan<byte> newData, ref int bytesDecoded, out string? line)
+            {
+                while (!newData.IsEmpty)
+                {
+                    // Buffer is small compared to a line.
+                    if (_length * 2 > _buffer!.Length)
+                    {
+                        char[] newBuffer = ArrayPool<char>.Shared.Rent(_length * 2);
+                        _buffer.AsSpan().Slice(_offset, _length).CopyTo(newBuffer);
+                        ArrayPool<char>.Shared.Return(_buffer);
+                        _buffer = newBuffer;
+                        _offset = 0;
+                    }
+                    // Data is past half of the buffer, move it to the front.
+                    if (_offset  * 2 > _buffer.Length)
+                    {
+                        _buffer.AsSpan().Slice(_offset, _length).CopyTo(_buffer);
+                        _offset = 0;
+                    }
+
+                    Span<char> remainder = _buffer.AsSpan(_offset + _length, _buffer!.Length - _offset - _length);
+                    _decoder!.Convert(newData, remainder, flush: false, out int bytesUsed, out int charsUsed, out bool completed);
```

Comment:
TODO: what does Decoder.Convert do when newData has more chars than what fit the remainder?